accepted/0000-robust-lifecycle-scripts.md

```diff
@@ -0,0 +1,71 @@
+# {{TITLE: a human-readable title for this RFC!}}
+
+## Summary
+
+Lifecycle scripts need to have more context passed to them about why an event is being invoked. Due to the way packages are deduped and hoisted, not all use cases are properly covered by simply event names.
+
+## Motivation
+
+Lifecycle scripts such as `preinstall` `install` and `postinstall` are very commonly used, but difficult to reason about due to the way package trees are installed and updated. In the case of `install`, a package could be directly installed as a dependency by a user, installed as a global, indirectly installed as a dependency, or added when dependency versions changed to the point where a dependency could no longer be shared between two packages as a single version. Worst still, `uninstall` could be fired when the package won't be removed at all due to one dependant no longer requiring it, but another still needing to keep it.
+
+An `uninstall` script could assume that the directory will be removed or not used in the future, and put the package into a bad state if it was merely deduped, for example. By providing scripts with more context, like whether a new directory is being made or destroyed, which depenants caused the installation, etc, we could properly cover all use-cases and remove confusion that would lead to destroyed data, bugs, and frustration.
+
+Future versions of npm could use different reify engines and methods, potentially leading to more needed context for these scripts.
+
+The primary reason for this RFC is the lack of `uninstall` lifecycle scripts in npm v7, and how difficult it would be to re-add them in a clear way.
+
+See issue: #3042
+
+## Detailed Explanation
+
+Lifecycle scripts need to be re-engineered and the old approach should be deprecated.
+
+`package.json` will include a new index called `"lifecycle"`
+
+```js
+{
+  "lifecycle": {
+    "install": {
```

Comment:
I think having discrete points in the reify process is better than trying to support the pre/post paradigm like this.  It was an inadequate model and confuses people a lot (i.e. "pre" means "it runs before the x scripts" not "it runs before x")

I'd prefer not having pre/post be baked in like this. If we have a need for something to run at a time, we put an event there and name it accordingly.

Reply:
imo "it runs before x" is the semantic every single person wants, and the other semantic is "silly sugar for `&&`".